### PR TITLE
[Tune] [PBT] Maintain consistent `Trial`/`TrialRunner` state when pausing and resuming trial

### PR DESCRIPTION
**Pull request by justinvyu**
### The problem
- When running *synchronous PBT* while checkpointing every time a perturbation happens, the experiment can reach a state where trial A is `RUNNING` but hanging forever without ever performing another train step, and trial B is `PAUSED` waiting for A to reach the specified `perturbation_interval`.

### Why does this happen?
1. Synch PBT waits for the last trial to come in to perform exploiting for all trials
2. PBT can call `TrialExecutor.stop_trial(trial)` within `PBT._exploit` before one of the other trials is finished saving (`trial.is_saving` is still True, and there is a decision in `TrialRunner._cached_trial_decisions` associated with this trial)
3. `TrialExecutor.stop_trial()` will clear all the futures that were to be handled by the trial (including `TrialRunner._process_trial_save` on the SAVING_RESULT event, which is the event that clears `trial.saving_to` and pops from `TrialRunner._cached_trial_decisions`)
4. This causes `trial.saving_to` to never be cleared, and `trial.is_saving` will remain True
5. Another training result event will come in due to `on_pg_ready` when the trial starts again (resuming from checkpoint)
    - When train → process_trial_result → goes into the `trial.is_saving` code path, which only adds the decision to the cache (without a SAVING_RESULT to move it to the decision queue) → trial is hanging forever
    - `TrialRunner._post_process_on_training_saving_result` will not do anything, since it checks that the trial is not in the `TrialRunner._cached_trial_decisions`
        - No actions will ever be executed

### Fix in the PR
- The main culprits here are inconsistent `Trial.saving_to`/`Trial.is_saving` and `TrialRunner._cached_trial_decisions`. These are now reset for the trial upon pausing.

### Testing
- This PR includes a test that reproduces this failure mode on the current `master` and is fixed with the PR. The test artificially creates the scenario by having one trial's checkpointing take a long time (5s), while PBT tries to pause that trial to exploit the other one.

### Future TODOs
- `PBT` directly calling `trial_runner.pause_trial(trial)` is not ideal to begin with, and it's the cause of this issue in the first place
- Refactor this in the future to clearly separate responsibilities between scheduler and trial runner/executor.
- Make sure that experiment restore is working when PBT pauses trials when other trials are checkpointing.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/execution/ray_trial_executor.py [hunk 1]
```diff
@@ -606,20 +606,22 @@ def stop_trial(
...
             for result_id in out:
                 self._futures.pop(result_id)
+        trial.saving_to = None
 
     def continue_training(self, trial: Trial) -> None:
         """Continues the training of this trial."""
         self._train(trial)
 
-    def pause_trial(self, trial: Trial) -> None:
+    def pause_trial(self, trial: Trial, should_checkpoint: bool = True) -> None:
```
Comment (justinvyu) [inline python/ray/tune/execution/ray_trial_executor.py:615]: Should I avoid having this be configurable? I think a lot of other places in the code depend on a paused trial always having an in-memory checkpoint. This one case of PBT will never use the trial's own in-memory checkpoint, which is why I would want this as False.

File: python/ray/tune/execution/trial_runner.py [hunk 2]
```diff
@@ -1295,7 +1295,7 @@ def _execute_action(self, trial: Trial, decision: str):
         if decision == TrialScheduler.CONTINUE:
             self.trial_executor.continue_training(trial)
         elif decision == TrialScheduler.PAUSE:
-            self.trial_executor.pause_trial(trial)
+            self.pause_trial(trial)
```
Comment (justinvyu) [inline python/ray/tune/execution/trial_runner.py:1298]: I made clearing the trial's entry in `_cached_trial_decisions` the default behavior for pausing.

File: python/ray/tune/execution/ray_trial_executor.py (hunk 1 above)
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py:618]: I think this is fine - it's the main reason why we call `stop` and then `set_status` in PBT at the moment, and I think this is cleaner.

File: python/ray/tune/execution/trial_runner.py [hunk 3]
```diff
@@ -977,7 +977,7 @@ def _on_training_result(self, trial, result):
     def _post_process_on_training_saving_result(self, trial):
         # `self._queued_trial_decisions` now contains a final decision
         # based on all results
-        if trial not in self._cached_trial_decisions:
+        if trial.trial_id not in self._cached_trial_decisions:
             final_decision = self._queued_trial_decisions.pop(trial.trial_id, None)
             if final_decision:
                 self._execute_action(trial, final_decision)
```
Comment (krfricke) [inline python/ray/tune/execution/trial_runner.py:983]: I actually think we can get rid if line 980 altogether (previously it was always True)
```suggestion
        final_decision = self._queued_trial_decisions.pop(trial.trial_id, None)
        if final_decision:
            self._execute_action(trial, final_decision)
```

**Review by krfricke — APPROVED** [review]
LGTM!

File: python/ray/tune/execution/ray_trial_executor.py (hunk 1 above)
Comment (xwjiang2010) [reply python/ray/tune/execution/ray_trial_executor.py:618]: Could we add some documentation here?

File: python/ray/tune/tests/test_trial_scheduler_pbt.py [hunk 4]
```diff
@@ -264,6 +267,102 @@ def testSynchPassLast(self):
             )
         )
 
+    def testExploitWhileSavingTrial(self):
```
Comment (xwjiang2010) [inline python/ray/tune/tests/test_trial_scheduler_pbt.py:270]: Thank you for adding this test case!

**Review by xwjiang2010 — APPROVED** [review]
Thank you Justin!

File: python/ray/tune/execution/trial_runner.py (hunk 2 above)
Comment (xwjiang2010) [reply python/ray/tune/execution/trial_runner.py:1297]: sounds good!

**Review by xwjiang2010 — APPROVED** [review]

**Review by krfricke — APPROVED** [review]
Awesome!

**Pull request merged**